File: 개념스터디/merge_sort.py
def merge_sort(arr):
    #devide
    if len(arr) < 2:
        return arr
    mid = len(arr) // 2
    left_arr = merge_sort(arr[:mid])
    right_arr = merge_sort(arr[mid:])
    print("분할", left_arr, right_arr)
    #merge
    merge_arr = []
    left = right = 0
    #현재
    while left < len(left_arr) and right < len(right_arr):
        if left_arr[left] < right_arr[right]:
            merge_arr.append(left_arr[left])
            left += 1
        else: #오른쪽이 더 작다면 
            merge_arr.append(right_arr[right])
            right += 1
  
    #남아 있는 값 더함 
    merge_arr += left_arr[left:]
    merge_arr += right_arr[right:]
    # print(left_arr[left:],left, right_arr[right:], right)
    return merge_arr

print(merge_sort([21,10,12,20,23,15,22]))

File: 개념스터디/Swift/merge_sort.swift
func mergeSort(_ arr: [Int]) -> [Int] {
    if arr.count < 2 {
        return arr
    }
    let mid = arr.count / 2
    let leftArr = mergeSort(Array(arr[0..<mid]))
    let rightArr = mergeSort(Array(arr[mid..<arr.count]))
    
    func merge(_ left: [Int], _ right: [Int]) -> [Int] {
        var left = left
        var right = right
        var result: [Int] = []
        
        while !left.isEmpty && !right.isEmpty { 
            //더 작은 값을 제거하면서 result에 더해줌 -> merge 단계
            if left[0] < right[0] {
                result.append(left.removeFirst())
            } else {
                result.append(right.removeFirst())
            }
        }
        
        if !left.isEmpty {
            result.append(contentsOf: left) //result에 남아있는 left 배열 더함
        } 
        if !right.isEmpty {
            result.append(contentsOf: right) 
        }
        return result
    }
    return merge(leftArr, rightArr)
}

print(mergeSort([21,10,12,20,23,15,22]))